{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d0550235",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Vagimock-and-gutmock:-obtain-desired-abundance-table\" data-toc-modified-id=\"Vagimock-and-gutmock:-obtain-desired-abundance-table-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Vagimock and gutmock: obtain desired abundance table</a></span><ul class=\"toc-item\"><li><span><a href=\"#Set-up\" data-toc-modified-id=\"Set-up-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Set up</a></span></li><li><span><a href=\"#Get-classifier\" data-toc-modified-id=\"Get-classifier-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Get classifier</a></span></li><li><span><a href=\"#Get-taxonomy\" data-toc-modified-id=\"Get-taxonomy-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Get taxonomy</a></span></li><li><span><a href=\"#Get-Taxa-Barplots\" data-toc-modified-id=\"Get-Taxa-Barplots-1.4\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>Get Taxa Barplots</a></span></li><li><span><a href=\"#Read-Taxa-Barplots\" data-toc-modified-id=\"Read-Taxa-Barplots-1.5\"><span class=\"toc-item-num\">1.5&nbsp;&nbsp;</span>Read Taxa Barplots</a></span></li><li><span><a href=\"#Obtain-Abundance-table\" data-toc-modified-id=\"Obtain-Abundance-table-1.6\"><span class=\"toc-item-num\">1.6&nbsp;&nbsp;</span>Obtain Abundance table</a></span></li></ul></li><li><span><a href=\"#Mockrobiota\" data-toc-modified-id=\"Mockrobiota-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Mockrobiota</a></span><ul class=\"toc-item\"><li><span><a href=\"#Download-data\" data-toc-modified-id=\"Download-data-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Download data</a></span></li><li><span><a href=\"#Filter-mockrobiota-records\" data-toc-modified-id=\"Filter-mockrobiota-records-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Filter mockrobiota records</a></span></li><li><span><a href=\"#Format-taxonomy-file\" data-toc-modified-id=\"Format-taxonomy-file-2.3\"><span class=\"toc-item-num\">2.3&nbsp;&nbsp;</span>Format taxonomy file</a></span></li><li><span><a href=\"#Create-abundance_env.txt\" data-toc-modified-id=\"Create-abundance_env.txt-2.4\"><span class=\"toc-item-num\">2.4&nbsp;&nbsp;</span>Create abundance_env.txt</a></span><ul class=\"toc-item\"><li><ul class=\"toc-item\"><li><span><a href=\"#Approach-2\" data-toc-modified-id=\"Approach-2-2.4.0.1\"><span class=\"toc-item-num\">2.4.0.1&nbsp;&nbsp;</span>Approach 2</a></span></li><li><span><a href=\"#Approach-1\" data-toc-modified-id=\"Approach-1-2.4.0.2\"><span class=\"toc-item-num\">2.4.0.2&nbsp;&nbsp;</span>Approach 1</a></span></li></ul></li></ul></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fcbd2a7",
   "metadata": {},
   "source": [
    "This notebook has been though to obtain essential files for the vagimock, gutmock and mockrobiota mock-communities. \n",
    "\n",
    "For **vagimock** and **gutmock** we will obtain the species that are suposed to be present in a specific sample environment such as the vagina or gut.\n",
    "- Input: abundancy table of previous analysis, containing taxon and frequency.\n",
    "- Ouput: expected-abundancy.tsv\n",
    "\n",
    "For **mockrobiota**, we will download and process data to unify taxonomies.\n",
    "- Input: itgdbmock\n",
    "- Output: fasta and taxa mockrobiota files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62870d40",
   "metadata": {},
   "source": [
    "# Vagimock and gutmock: obtain desired abundance table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ec9057d",
   "metadata": {},
   "source": [
    "## Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d1e941d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from qiime2 import Artifact\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from IPython.display import display\n",
    "import os\n",
    "from os.path import join, basename, dirname, exists, splitext\n",
    "\n",
    "from joblib import Parallel, delayed\n",
    "import random\n",
    "\n",
    "%matplotlib inline\n",
    "%load_ext autoreload\n",
    "\n",
    "# Custom functions\n",
    "import mock_community\n",
    "from utils import check_dir\n",
    "import utils\n",
    "import db_comparison\n",
    "\n",
    "import logging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "473cac13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mock-communities\n",
    "mock_names = ['vagimock', 'gutmock']\n",
    "\n",
    "# Working directory\n",
    "pwd = os.getcwd()\n",
    "\n",
    "# Database files\n",
    "created_dir = \"new_approach/created_db\"\n",
    "original_dir = \"new_approach/original_db\"\n",
    "\n",
    "db = {'gsrv_V4':\n",
    "      {'seq':join(created_dir, 'gsrv_V4_seqs.qza'),\n",
    "      'taxa':join(created_dir, 'gsrv_V4_taxa.qza')}\n",
    "}\n",
    "\n",
    "## Database parameters\n",
    "db_dir = created_dir\n",
    "db_name = 'gsrv_V4'\n",
    "db_version = \"gsrv_V4\"\n",
    "db_similarity = \"99-otus\"\n",
    "\n",
    "## Database output files\n",
    "db_out_files = {\n",
    "    mocknames: join(mocknames, 'abundance_table'\n",
    "                ) for mocknames in mock_names}\n",
    "\n",
    "\n",
    "# Data input directory: Qiime2 files (taxonomy and table)\n",
    "in_dirs = [\"/media/VMP\",\n",
    "          \"/media/LT\"]\n",
    "data_dir = { mockname: d for mockname, d in zip(mock_names, in_dirs)}\n",
    "\n",
    "# Inputs\n",
    "## representative sequences of sample specific environment (Rep-seqs)\n",
    "rep_seqs_path = [\n",
    "    \"rep-seqs_merged_VMP_run29_40_46.qza\", \n",
    "    \"rep-seqs-dada2_mergedRuns_LT.qza\" ]\n",
    "rep_seqs = {\n",
    "    mockname: join(\n",
    "        data_dir[mockname], rep_seq_path\n",
    "    ) for mockname, rep_seq_path in zip(mock_names, rep_seqs_path)}\n",
    "\n",
    "## Table abundancies of Rep-Seqs\n",
    "tab_paths = [\n",
    "    \"table_merged_VMP_run29_40_46.qza\", \n",
    "    \"table-dada2_mergedRuns_LT.qza\"]\n",
    "tabs = {\n",
    "    mockname: join(\n",
    "        data_dir[mockname], tab_path\n",
    "    ) for mockname, tab_path in zip(mock_names, tab_paths)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee038b70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate directory\n",
    "for key, outdir in db_out_files.items():\n",
    "    check_dir(outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba212e2d",
   "metadata": {},
   "source": [
    "We will obtain the abundance table from a previous analyzed data. Therefore, we will generate the classifier, perform the taxonomy assingment and obtain the taxa-barplots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f579ba8",
   "metadata": {},
   "source": [
    "## Get classifier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f045daad",
   "metadata": {},
   "source": [
    "Build the classifier with default pgram ([7,7])."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc44196d",
   "metadata": {},
   "outputs": [],
   "source": [
    "classifier_dir = 'db_evaluation/classifiers'\n",
    "\n",
    "%autoreload\n",
    "cmds, classifier_path = db_comparison.get_classifier(\n",
    "    db, classifier_dir, bespoke=False, p_alpha=0.001, p_feat='[7,7]', \n",
    "    force=True)\n",
    "cmds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20bafe66",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "Parallel(n_jobs=38)(delayed(os.system)(cmd) for cmd in cmds);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a5d5cb3",
   "metadata": {},
   "source": [
    "## Get taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa05ad7b",
   "metadata": {},
   "source": [
    "Perform the taxonomy assignment with default confidence 0.7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fb298106",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 1s, sys: 11 s, total: 1min 12s\n",
      "Wall time: 5min 47s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "all_tax = []\n",
    "\n",
    "for mockname, path in rep_seqs.items():\n",
    "    rs_artifact = Artifact.load(path)\n",
    "    tax = db_comparison.get_taxonomy(\n",
    "            rep_seqs_artifact = rs_artifact, \n",
    "            classifier_paths = classifier_path, \n",
    "            out_dir = pwd, label = mockname, method=\"naive-bayes\", \n",
    "            threads=35, mock_community=False, p_confidence=0.7, \n",
    "            force=True )\n",
    "    all_tax.append(tax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03d01e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_tax"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0e29a3f",
   "metadata": {},
   "source": [
    "## Get Taxa Barplots\n",
    "We will collapse the frequency table at species level, obtaining the abundancy at specie level for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "584524b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%autoreload\n",
    "tax_barplots = []\n",
    "\n",
    "for i , tax_dict in enumerate(all_tax):\n",
    "    tax = list(tax_dict.values())[0][0]\n",
    "    mockname = mock_names[i]\n",
    "    tab = Artifact.load(tabs[mockname])\n",
    "    path = join(\n",
    "        db_out_files[mockname], \n",
    "        f\"taxa_barplot-L7_{db_name}_0.001::[7,7]:0.7_{mockname}\")\n",
    "    \n",
    "    tsv_tab = mock_community.generate_taxa_barplots(\n",
    "        tax_artifact = tax, tab_artifact = tab,\n",
    "        level = 7, output = f\"{path}.tsv\" , force = True\n",
    "    \n",
    "    )\n",
    "    \n",
    "    tax_barplots.append(tsv_tab)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f473ef98",
   "metadata": {},
   "source": [
    "## Read Taxa Barplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ea6f8751",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Taxa-Barplots\n",
    "tb_df = {}\n",
    "for tb in tax_barplots:\n",
    "    key = tb.split('/')[0]\n",
    "    df = pd.read_csv(tb, header = 1, index_col = 0, sep='\\t')\n",
    "    \n",
    "    tb_df[key] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0c1ade46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#OTU ID</th>\n",
       "      <th>INT.CTL_Run40</th>\n",
       "      <th>V.SWAB.13CE.1</th>\n",
       "      <th>V.SWAB.13CE.2</th>\n",
       "      <th>V.SWAB.15CE.1</th>\n",
       "      <th>V.SWAB.15CE.2</th>\n",
       "      <th>V.SWAB.16C.1.RUN29</th>\n",
       "      <th>V.SWAB.20CE.1</th>\n",
       "      <th>V.SWAB.20CE.2</th>\n",
       "      <th>V.SWAB.21CE.1</th>\n",
       "      <th>...</th>\n",
       "      <th>PBV.70.1</th>\n",
       "      <th>PBV.72.1</th>\n",
       "      <th>PBV.72.2</th>\n",
       "      <th>PBV.73.1</th>\n",
       "      <th>PBV.75.1</th>\n",
       "      <th>PBV.77.1</th>\n",
       "      <th>PBV.77.2</th>\n",
       "      <th>V.SWAB.21PE.2_R</th>\n",
       "      <th>V.SWAB.3.CE.2_R</th>\n",
       "      <th>V.SWAB.REUS2.2_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria;__;__;__;__;__;__</td>\n",
       "      <td>313.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria;p__Firmicutes;c__Clostridia;o__Eub...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...</td>\n",
       "      <td>334.0</td>\n",
       "      <td>1262.0</td>\n",
       "      <td>961.0</td>\n",
       "      <td>105772.0</td>\n",
       "      <td>86887.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>333.0</td>\n",
       "      <td>1021.0</td>\n",
       "      <td>531.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59495.0</td>\n",
       "      <td>12205.0</td>\n",
       "      <td>2841.0</td>\n",
       "      <td>359.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>9651.0</td>\n",
       "      <td>531.0</td>\n",
       "      <td>326.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria;p__Firmicutes;c__Tissierellia;o__;...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria;p__Firmicutes;c__Tissierellia;o__T...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>174.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>k__Bacteria;p__Firmicutes;c__Clostridia;o__Eub...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>k__Bacteria;p__Bacteroidetes;c__Saprospiria;o_...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>k__Bacteria;p__Firmicutes;c__Clostridia;o__Eub...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>k__Bacteria;p__Actinobacteria;c__Coriobacterii...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>589</th>\n",
       "      <td>k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...</td>\n",
       "      <td>41.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>590 rows × 259 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               #OTU ID  INT.CTL_Run40  \\\n",
       "0                        k__Bacteria;__;__;__;__;__;__          313.0   \n",
       "1    k__Bacteria;p__Firmicutes;c__Clostridia;o__Eub...            0.0   \n",
       "2    k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...          334.0   \n",
       "3    k__Bacteria;p__Firmicutes;c__Tissierellia;o__;...            0.0   \n",
       "4    k__Bacteria;p__Firmicutes;c__Tissierellia;o__T...            0.0   \n",
       "..                                                 ...            ...   \n",
       "585  k__Bacteria;p__Firmicutes;c__Clostridia;o__Eub...            0.0   \n",
       "586  k__Bacteria;p__Bacteroidetes;c__Saprospiria;o_...            0.0   \n",
       "587  k__Bacteria;p__Firmicutes;c__Clostridia;o__Eub...           44.0   \n",
       "588  k__Bacteria;p__Actinobacteria;c__Coriobacterii...            0.0   \n",
       "589  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...           41.0   \n",
       "\n",
       "     V.SWAB.13CE.1  V.SWAB.13CE.2  V.SWAB.15CE.1  V.SWAB.15CE.2  \\\n",
       "0              0.0            0.0            0.0            0.0   \n",
       "1              0.0            0.0            0.0            0.0   \n",
       "2           1262.0          961.0       105772.0        86887.0   \n",
       "3              0.0          100.0            0.0            0.0   \n",
       "4              0.0          169.0            0.0            0.0   \n",
       "..             ...            ...            ...            ...   \n",
       "585            0.0            0.0            0.0            0.0   \n",
       "586            0.0            0.0            0.0            0.0   \n",
       "587            0.0            0.0            0.0            0.0   \n",
       "588            0.0            0.0            0.0            0.0   \n",
       "589            0.0            0.0            0.0            0.0   \n",
       "\n",
       "     V.SWAB.16C.1.RUN29  V.SWAB.20CE.1  V.SWAB.20CE.2  V.SWAB.21CE.1  ...  \\\n",
       "0                   0.0            0.0            0.0            0.0  ...   \n",
       "1                   0.0            0.0            0.0            0.0  ...   \n",
       "2                 221.0          333.0         1021.0          531.0  ...   \n",
       "3                   0.0            0.0            0.0            0.0  ...   \n",
       "4                   0.0           13.0            0.0            0.0  ...   \n",
       "..                  ...            ...            ...            ...  ...   \n",
       "585                 0.0            0.0            0.0            0.0  ...   \n",
       "586                 0.0            0.0            0.0            0.0  ...   \n",
       "587                 0.0            0.0            0.0            0.0  ...   \n",
       "588                 0.0            0.0            0.0            0.0  ...   \n",
       "589                 0.0            0.0            0.0            0.0  ...   \n",
       "\n",
       "     PBV.70.1  PBV.72.1  PBV.72.2  PBV.73.1  PBV.75.1  PBV.77.1  PBV.77.2  \\\n",
       "0         0.0       4.0       0.0       0.0       0.0       0.0      13.0   \n",
       "1         0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2     59495.0   12205.0    2841.0     359.0     116.0     150.0     171.0   \n",
       "3         0.0       0.0       0.0      31.0      80.0       0.0       0.0   \n",
       "4       174.0       0.0       0.0      34.0     390.0       0.0       0.0   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "585       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "586       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "587       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "588       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "589       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "     V.SWAB.21PE.2_R  V.SWAB.3.CE.2_R  V.SWAB.REUS2.2_R  \n",
       "0                0.0              0.0              10.0  \n",
       "1                0.0              0.0               0.0  \n",
       "2             9651.0            531.0             326.0  \n",
       "3                0.0              0.0               0.0  \n",
       "4                0.0              0.0               0.0  \n",
       "..               ...              ...               ...  \n",
       "585              0.0              0.0               0.0  \n",
       "586              0.0              0.0               0.0  \n",
       "587              0.0              0.0               0.0  \n",
       "588              0.0              0.0               0.0  \n",
       "589              0.0              0.0               0.0  \n",
       "\n",
       "[590 rows x 259 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tb_df['vagimock'].reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd91cbdd",
   "metadata": {},
   "source": [
    "Now, we will modified this QIIME table so that the taxonomy syntax matches the syntax database. Specifically:\n",
    "\n",
    "- we have add a space between the separator (';') of the clades\n",
    "- add the prefix (e.g. s, g...) before an unspecified clade\n",
    "\n",
    "Only keep rows with identified species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "29980785",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>INT.CTL_Run40</th>\n",
       "      <th>V.SWAB.13CE.1</th>\n",
       "      <th>V.SWAB.13CE.2</th>\n",
       "      <th>V.SWAB.15CE.1</th>\n",
       "      <th>V.SWAB.15CE.2</th>\n",
       "      <th>V.SWAB.16C.1.RUN29</th>\n",
       "      <th>V.SWAB.20CE.1</th>\n",
       "      <th>V.SWAB.20CE.2</th>\n",
       "      <th>V.SWAB.21CE.1</th>\n",
       "      <th>...</th>\n",
       "      <th>PBV.70.1</th>\n",
       "      <th>PBV.72.1</th>\n",
       "      <th>PBV.72.2</th>\n",
       "      <th>PBV.73.1</th>\n",
       "      <th>PBV.75.1</th>\n",
       "      <th>PBV.77.1</th>\n",
       "      <th>PBV.77.2</th>\n",
       "      <th>V.SWAB.21PE.2_R</th>\n",
       "      <th>V.SWAB.3.CE.2_R</th>\n",
       "      <th>V.SWAB.REUS2.2_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria; p__Actinobacteria; c__Coriobacter...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Negativicutes; ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>441.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>558.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>740.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>400</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Saprospiria;...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>k__Bacteria; p__Actinobacteria; c__Coriobacter...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>41.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>404 rows × 259 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Taxon  INT.CTL_Run40  \\\n",
       "0    k__Bacteria; p__Firmicutes; c__Clostridia; o__...            0.0   \n",
       "1    k__Bacteria; p__Actinobacteria; c__Coriobacter...            0.0   \n",
       "2    k__Bacteria; p__Firmicutes; c__Negativicutes; ...            0.0   \n",
       "3    k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...            0.0   \n",
       "4    k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...          740.0   \n",
       "..                                                 ...            ...   \n",
       "399  k__Bacteria; p__Firmicutes; c__Clostridia; o__...            0.0   \n",
       "400  k__Bacteria; p__Firmicutes; c__Clostridia; o__...            0.0   \n",
       "401  k__Bacteria; p__Bacteroidetes; c__Saprospiria;...            0.0   \n",
       "402  k__Bacteria; p__Actinobacteria; c__Coriobacter...            0.0   \n",
       "403  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...           41.0   \n",
       "\n",
       "     V.SWAB.13CE.1  V.SWAB.13CE.2  V.SWAB.15CE.1  V.SWAB.15CE.2  \\\n",
       "0              0.0            0.0            0.0            0.0   \n",
       "1              0.0            0.0            0.0            0.0   \n",
       "2              0.0            0.0            0.0            7.0   \n",
       "3              0.0            0.0            0.0            0.0   \n",
       "4              0.0            0.0            0.0            0.0   \n",
       "..             ...            ...            ...            ...   \n",
       "399            0.0            0.0            0.0            0.0   \n",
       "400            0.0            0.0            0.0            0.0   \n",
       "401            0.0            0.0            0.0            0.0   \n",
       "402            0.0            0.0            0.0            0.0   \n",
       "403            0.0            0.0            0.0            0.0   \n",
       "\n",
       "     V.SWAB.16C.1.RUN29  V.SWAB.20CE.1  V.SWAB.20CE.2  V.SWAB.21CE.1  ...  \\\n",
       "0                   0.0            0.0            0.0            0.0  ...   \n",
       "1                   0.0            0.0            0.0            0.0  ...   \n",
       "2                   0.0            0.0          142.0            0.0  ...   \n",
       "3                   0.0            0.0          441.0            0.0  ...   \n",
       "4                   0.0            0.0            0.0            0.0  ...   \n",
       "..                  ...            ...            ...            ...  ...   \n",
       "399                 0.0            0.0            0.0            0.0  ...   \n",
       "400                 0.0            0.0            0.0            0.0  ...   \n",
       "401                 0.0            0.0            0.0            0.0  ...   \n",
       "402                 0.0            0.0            0.0            0.0  ...   \n",
       "403                 0.0            0.0            0.0            0.0  ...   \n",
       "\n",
       "     PBV.70.1  PBV.72.1  PBV.72.2  PBV.73.1  PBV.75.1  PBV.77.1  PBV.77.2  \\\n",
       "0         0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "1         0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2         0.0       0.0       0.0       0.0     103.0       0.0       0.0   \n",
       "3         0.0       0.0       0.0       0.0     558.0       0.0       0.0   \n",
       "4         0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "399       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "400       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "401       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "402       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "403       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "     V.SWAB.21PE.2_R  V.SWAB.3.CE.2_R  V.SWAB.REUS2.2_R  \n",
       "0                0.0              0.0               0.0  \n",
       "1                0.0              0.0               0.0  \n",
       "2                0.0              0.0               0.0  \n",
       "3                0.0              0.0               0.0  \n",
       "4                0.0              0.0               0.0  \n",
       "..               ...              ...               ...  \n",
       "399              0.0              0.0               0.0  \n",
       "400              0.0              0.0               0.0  \n",
       "401              0.0              0.0               0.0  \n",
       "402              0.0              0.0               0.0  \n",
       "403              0.0              0.0               0.0  \n",
       "\n",
       "[404 rows x 259 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>D001_LT</th>\n",
       "      <th>D002_LT</th>\n",
       "      <th>D003_LT</th>\n",
       "      <th>D004_LT</th>\n",
       "      <th>D005_LT</th>\n",
       "      <th>HR002_LT</th>\n",
       "      <th>HR004_LT</th>\n",
       "      <th>HR005_LT</th>\n",
       "      <th>HR007_LT</th>\n",
       "      <th>...</th>\n",
       "      <th>P023.6_LT</th>\n",
       "      <th>P023.7.m2_LT</th>\n",
       "      <th>P023.8_LT</th>\n",
       "      <th>P023.9_LT</th>\n",
       "      <th>P024.3_LT</th>\n",
       "      <th>P024.5_LT</th>\n",
       "      <th>P024.6_LT</th>\n",
       "      <th>P024.7_LT</th>\n",
       "      <th>P024.8_LT</th>\n",
       "      <th>P024.9_LT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>261.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>138.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>271.0</td>\n",
       "      <td>182.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>416.0</td>\n",
       "      <td>784.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1914.0</td>\n",
       "      <td>206.0</td>\n",
       "      <td>527.0</td>\n",
       "      <td>598.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Erysipelotrichi...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>180.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria; p__Proteobacteria; c__Betaproteob...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria; p__Actinobacteria; c__Actinomycet...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>116.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>k__Bacteria; p__Proteobacteria; c__Alphaproteo...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>k__Archaea; p__Euryarchaeota; c__Methanomicrob...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>437 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Taxon  D001_LT  D002_LT  \\\n",
       "0    k__Bacteria; p__Firmicutes; c__Clostridia; o__...    261.0      0.0   \n",
       "1    k__Bacteria; p__Firmicutes; c__Clostridia; o__...      0.0     66.0   \n",
       "2    k__Bacteria; p__Firmicutes; c__Erysipelotrichi...      0.0      0.0   \n",
       "3    k__Bacteria; p__Proteobacteria; c__Betaproteob...      0.0    159.0   \n",
       "4    k__Bacteria; p__Actinobacteria; c__Actinomycet...      0.0      0.0   \n",
       "..                                                 ...      ...      ...   \n",
       "432  k__Bacteria; p__Proteobacteria; c__Alphaproteo...      0.0      0.0   \n",
       "433  k__Bacteria; p__Firmicutes; c__Clostridia; o__...      0.0      0.0   \n",
       "434  k__Archaea; p__Euryarchaeota; c__Methanomicrob...      0.0      0.0   \n",
       "435  k__Bacteria; p__Firmicutes; c__Clostridia; o__...      0.0      0.0   \n",
       "436  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...      0.0      0.0   \n",
       "\n",
       "     D003_LT  D004_LT  D005_LT  HR002_LT  HR004_LT  HR005_LT  HR007_LT  ...  \\\n",
       "0      287.0    138.0      0.0       0.0       0.0     271.0     182.0  ...   \n",
       "1        9.0      0.0      0.0       0.0     129.0     110.0       0.0  ...   \n",
       "2        9.0      0.0      0.0     103.0       0.0       0.0       0.0  ...   \n",
       "3        0.0      0.0      0.0       0.0       0.0      23.0      17.0  ...   \n",
       "4        0.0      0.0      0.0       0.0       0.0       0.0       0.0  ...   \n",
       "..       ...      ...      ...       ...       ...       ...       ...  ...   \n",
       "432      0.0      0.0      0.0       0.0       0.0       0.0       0.0  ...   \n",
       "433      0.0      0.0      0.0       0.0       0.0       0.0       0.0  ...   \n",
       "434      0.0      0.0      0.0       0.0       0.0       0.0       0.0  ...   \n",
       "435      0.0      0.0      0.0       0.0       0.0       0.0       0.0  ...   \n",
       "436      0.0      0.0      0.0       0.0       0.0       0.0       0.0  ...   \n",
       "\n",
       "     P023.6_LT  P023.7.m2_LT  P023.8_LT  P023.9_LT  P024.3_LT  P024.5_LT  \\\n",
       "0          0.0           0.0      416.0      784.0      135.0        0.0   \n",
       "1          0.0           0.0       15.0        0.0        0.0        0.0   \n",
       "2          0.0          91.0      172.0      110.0      107.0      148.0   \n",
       "3          0.0           0.0        0.0        0.0        0.0        0.0   \n",
       "4        116.0           5.0       13.0        9.0        0.0        0.0   \n",
       "..         ...           ...        ...        ...        ...        ...   \n",
       "432        0.0           2.0        0.0        0.0        0.0        0.0   \n",
       "433        0.0           0.0        0.0        0.0        0.0        0.0   \n",
       "434        0.0           0.0        0.0        0.0        0.0        0.0   \n",
       "435        0.0           0.0        0.0        0.0        0.0        0.0   \n",
       "436        0.0           0.0        0.0        0.0        0.0        0.0   \n",
       "\n",
       "     P024.6_LT  P024.7_LT  P024.8_LT  P024.9_LT  \n",
       "0       1914.0      206.0      527.0      598.0  \n",
       "1          0.0        0.0        0.0        0.0  \n",
       "2         92.0       85.0      133.0      180.0  \n",
       "3          0.0        0.0        0.0        0.0  \n",
       "4          0.0        0.0        0.0        0.0  \n",
       "..         ...        ...        ...        ...  \n",
       "432        0.0        0.0        0.0        0.0  \n",
       "433        0.0        0.0        0.0        0.0  \n",
       "434        0.0        0.0        0.0        0.0  \n",
       "435        0.0        0.0        0.0        0.0  \n",
       "436        0.0        0.0        0.0        0.0  \n",
       "\n",
       "[437 rows x 149 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key, df in tb_df.items():\n",
    "    # modify taxonomy to match the databases\n",
    "    df = utils.format_taxonomy_of_taxabarplots(df)\n",
    "    # Filter unindentified species\n",
    "    species = [\n",
    "        item for item in list(df['Taxon']) if not item.endswith('s__')]\n",
    "    df = df[df['Taxon'].isin(species)].reset_index(drop=True)\n",
    "    display(df)\n",
    "    tb_df[key] = df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4fb9132",
   "metadata": {},
   "source": [
    "## Obtain Abundance table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3f5df66",
   "metadata": {},
   "source": [
    "Split the dataframes in five groups. In order to include *Lactobacillus crispatus* in the vagimock community, we will increase its abundancy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f8edc092",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of groups\n",
    "i = 5 \n",
    "random.seed(123)\n",
    "\n",
    "for key, df in tb_df.items():\n",
    "    df = df.set_index('Taxon')\n",
    "    colnames = list(df.columns)\n",
    "    random.shuffle(colnames)\n",
    "    grouped_colnames = [colnames[x::i] for x in range(i)]\n",
    "    sub_dfs = [df[group].reset_index() for group in grouped_colnames]\n",
    "    \n",
    "    tb_df[key] = sub_dfs\n",
    "    \n",
    "# increase the abundancy of crispatus differently in each group:\n",
    "\n",
    "max_abun = [1, 1000, 10000, 50000, 100000]\n",
    "\n",
    "for i, (abun, df) in enumerate(zip(max_abun, tb_df[\"vagimock\"])):\n",
    "    df = df.set_index('Taxon')\n",
    "#     crispatus_row = df[df['Taxon'].str.contains('crispatus')]\n",
    "    crispatus_row = df.filter(like='crispatus', axis=0)\n",
    "    ncols = len(crispatus_row.columns)\n",
    "    crispatus_mod_row  = crispatus_row+random.choices(range(0,abun), k=ncols)\n",
    "    df.loc[crispatus_row.index, :] = crispatus_mod_row\n",
    "    df = df.reset_index()\n",
    "    tb_df['vagimock'][i] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4bd292d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>V.SWAB.SPAU2.2</th>\n",
       "      <th>PBV.4.1</th>\n",
       "      <th>V.SWAB.REUS1.1</th>\n",
       "      <th>V.SWAB.SPAU5.1</th>\n",
       "      <th>PBV.58.2</th>\n",
       "      <th>V.SWAB.21C.1</th>\n",
       "      <th>PBV.16.2</th>\n",
       "      <th>V.SWAB.14C.1</th>\n",
       "      <th>PBV.66.2</th>\n",
       "      <th>...</th>\n",
       "      <th>V.SWAB.15C.2</th>\n",
       "      <th>V.SWAB.53PE.2</th>\n",
       "      <th>V.SWAB.16C.2</th>\n",
       "      <th>V.SWAB.10PE.2</th>\n",
       "      <th>V.SWAB.15PE.1</th>\n",
       "      <th>PBV.64.2</th>\n",
       "      <th>V.SWAB.24CE.2</th>\n",
       "      <th>V.SWAB.3.CE.2_R</th>\n",
       "      <th>PBV.44.2</th>\n",
       "      <th>V.SWAB.HUMIC11.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria; p__Actinobacteria; c__Coriobacter...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Negativicutes; ...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Taxon  V.SWAB.SPAU2.2  PBV.4.1  \\\n",
       "0  k__Bacteria; p__Firmicutes; c__Clostridia; o__...             0.0      0.0   \n",
       "1  k__Bacteria; p__Actinobacteria; c__Coriobacter...             0.0      0.0   \n",
       "2  k__Bacteria; p__Firmicutes; c__Negativicutes; ...             0.0      0.0   \n",
       "3  k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...             0.0      0.0   \n",
       "4  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...             0.0      0.0   \n",
       "\n",
       "   V.SWAB.REUS1.1  V.SWAB.SPAU5.1  PBV.58.2  V.SWAB.21C.1  PBV.16.2  \\\n",
       "0             0.0             0.0       0.0           0.0       0.0   \n",
       "1             0.0             0.0       0.0           0.0       0.0   \n",
       "2             0.0             0.0       0.0           0.0       0.0   \n",
       "3             0.0             0.0       0.0           0.0       0.0   \n",
       "4             0.0             0.0       0.0           0.0       0.0   \n",
       "\n",
       "   V.SWAB.14C.1  PBV.66.2  ...  V.SWAB.15C.2  V.SWAB.53PE.2  V.SWAB.16C.2  \\\n",
       "0           0.0       0.0  ...           0.0            0.0           0.0   \n",
       "1           0.0       0.0  ...           0.0            0.0           0.0   \n",
       "2           0.0       0.0  ...           0.0            0.0           0.0   \n",
       "3           0.0       0.0  ...           0.0           45.0           0.0   \n",
       "4           0.0       0.0  ...           0.0            6.0           0.0   \n",
       "\n",
       "   V.SWAB.10PE.2  V.SWAB.15PE.1  PBV.64.2  V.SWAB.24CE.2  V.SWAB.3.CE.2_R  \\\n",
       "0            0.0            0.0       0.0            0.0              0.0   \n",
       "1            0.0            0.0       0.0           71.0              0.0   \n",
       "2            0.0            0.0       0.0            0.0              0.0   \n",
       "3            0.0            0.0       0.0            0.0              0.0   \n",
       "4            0.0            0.0       0.0            0.0              0.0   \n",
       "\n",
       "   PBV.44.2  V.SWAB.HUMIC11.1  \n",
       "0       0.0               0.0  \n",
       "1       0.0               0.0  \n",
       "2       0.0               0.0  \n",
       "3       0.0               0.0  \n",
       "4       0.0               0.0  \n",
       "\n",
       "[5 rows x 53 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(tb_df['vagimock'][0].head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bd94066",
   "metadata": {},
   "source": [
    "In each dataframe, keep the species that represents the 99% of the abundance. \n",
    "\n",
    "Obtain the relative abundancies. Note that the frequencies are multiplied by 100 in order to avoid very low numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7071ca51",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, items in tb_df.items():\n",
    "    for i, df in enumerate(items):\n",
    "        \n",
    "        # Calculate Total Abundance, Relative Abundance and  Rel Acumulative Abundance\n",
    "        df['Abundance'] = df.sum(axis=1)\n",
    "        df['Rel_Abundance'] = (df['Abundance']/df['Abundance'].sum())*100\n",
    "        df = df.sort_values(by=['Rel_Abundance'], ascending=False)\n",
    "        df['Rel_Abundance_cumulative'] = df['Rel_Abundance'].cumsum()\n",
    "        display(df)\n",
    "        # Only keep the species that represents the 99% of the abundance \n",
    "        df = df[df['Rel_Abundance_cumulative'] < 99]\n",
    "        \n",
    "        tb_df[key][i] = df \n",
    "        display(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "13b4e798",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>vagimock00</th>\n",
       "      <th>vagimock01</th>\n",
       "      <th>vagimock02</th>\n",
       "      <th>vagimock03</th>\n",
       "      <th>vagimock04</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...</td>\n",
       "      <td>92.858135</td>\n",
       "      <td>81.513449</td>\n",
       "      <td>68.090323</td>\n",
       "      <td>40.652198</td>\n",
       "      <td>22.501148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...</td>\n",
       "      <td>2.685409</td>\n",
       "      <td>5.038885</td>\n",
       "      <td>3.810807</td>\n",
       "      <td>3.009378</td>\n",
       "      <td>0.937033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria; p__Actinobacteria; c__Actinomycet...</td>\n",
       "      <td>1.385344</td>\n",
       "      <td>0.415924</td>\n",
       "      <td>0.852260</td>\n",
       "      <td>0.783070</td>\n",
       "      <td>0.957093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria; p__Actinobacteria; c__Actinomycet...</td>\n",
       "      <td>1.300811</td>\n",
       "      <td>4.166379</td>\n",
       "      <td>9.575885</td>\n",
       "      <td>6.888249</td>\n",
       "      <td>6.508824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>0.604868</td>\n",
       "      <td>0.122824</td>\n",
       "      <td>0.146558</td>\n",
       "      <td>0.087941</td>\n",
       "      <td>0.295946</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Taxon  vagimock00  vagimock01  \\\n",
       "0  k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...   92.858135   81.513449   \n",
       "1  k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...    2.685409    5.038885   \n",
       "2  k__Bacteria; p__Actinobacteria; c__Actinomycet...    1.385344    0.415924   \n",
       "3  k__Bacteria; p__Actinobacteria; c__Actinomycet...    1.300811    4.166379   \n",
       "4  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...    0.604868    0.122824   \n",
       "\n",
       "   vagimock02  vagimock03  vagimock04  \n",
       "0   68.090323   40.652198   22.501148  \n",
       "1    3.810807    3.009378    0.937033  \n",
       "2    0.852260    0.783070    0.957093  \n",
       "3    9.575885    6.888249    6.508824  \n",
       "4    0.146558    0.087941    0.295946  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>gutmock00</th>\n",
       "      <th>gutmock01</th>\n",
       "      <th>gutmock02</th>\n",
       "      <th>gutmock03</th>\n",
       "      <th>gutmock04</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>11.777871</td>\n",
       "      <td>11.084972</td>\n",
       "      <td>16.278256</td>\n",
       "      <td>9.095578</td>\n",
       "      <td>16.141840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>5.551624</td>\n",
       "      <td>0.577970</td>\n",
       "      <td>0.890686</td>\n",
       "      <td>0.690637</td>\n",
       "      <td>0.412421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>4.428748</td>\n",
       "      <td>6.483817</td>\n",
       "      <td>5.711091</td>\n",
       "      <td>5.769315</td>\n",
       "      <td>4.541067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>3.812079</td>\n",
       "      <td>5.267262</td>\n",
       "      <td>4.116979</td>\n",
       "      <td>2.088251</td>\n",
       "      <td>8.776298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>3.662234</td>\n",
       "      <td>4.626383</td>\n",
       "      <td>2.786266</td>\n",
       "      <td>3.276952</td>\n",
       "      <td>1.875877</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Taxon  gutmock00  gutmock01  \\\n",
       "0  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...  11.777871  11.084972   \n",
       "1  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...   5.551624   0.577970   \n",
       "2  k__Bacteria; p__Firmicutes; c__Clostridia; o__...   4.428748   6.483817   \n",
       "3  k__Bacteria; p__Firmicutes; c__Clostridia; o__...   3.812079   5.267262   \n",
       "4  k__Bacteria; p__Firmicutes; c__Clostridia; o__...   3.662234   4.626383   \n",
       "\n",
       "   gutmock02  gutmock03  gutmock04  \n",
       "0  16.278256   9.095578  16.141840  \n",
       "1   0.890686   0.690637   0.412421  \n",
       "2   5.711091   5.769315   4.541067  \n",
       "3   4.116979   2.088251   8.776298  \n",
       "4   2.786266   3.276952   1.875877  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "abundance_tabs = {}\n",
    "for key, items in tb_df.items():\n",
    "    d = {}\n",
    "    \n",
    "    for i, item in enumerate(items):\n",
    "        name = f\"{key}0{i}\"\n",
    "        d[name] = {}\n",
    "        for j in item.index:\n",
    "            tax = str(item.loc[j, 'Taxon'])\n",
    "            d[name][tax] = item.loc[j, 'Rel_Abundance']\n",
    "        \n",
    "    \n",
    "    df = pd.DataFrame.from_dict(d)\n",
    "    df = df.fillna(0)\n",
    "    df.reset_index(inplace=True)\n",
    "    df = df.rename(columns = {'index': 'Taxon'})\n",
    "    abundance_tabs[key] = df\n",
    "    \n",
    "    display(df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d195d77",
   "metadata": {},
   "source": [
    "The above abundance tables contain the expected abundances for each taxon. We will save this abundancy table as in tsv."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8666fc93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the table\n",
    "force=True\n",
    "\n",
    "for key, tab in abundance_tabs.items():\n",
    "    exp_tax_out = join(db_out_files[key],\"expected-taxonomy.tsv\")\n",
    "    \n",
    "    if not exists(exp_tax_out) or force:\n",
    "        tab.to_csv(exp_tax_out, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e46f9828",
   "metadata": {},
   "source": [
    "# Mockrobiota"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b738ebe",
   "metadata": {},
   "source": [
    "## Download data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe0d1d86",
   "metadata": {},
   "source": [
    "The data was obtained from 16S-itgdb paper. [Link](https://github.com/yphsieh/16S-ITGDB/tree/master/test%20cases) . We downloaded the combined mock validation dataset, containing PacBio HMP, Zymo and Mockrobiota mock communities:\n",
    "- [final_mock_seq.fasta](https://drive.google.com/file/d/1Y4PeUcZAuXkB2uU0OBv7uygqQlX2tXXp/view?usp=share_link)\n",
    "- [final_mock_taxa.txt](https://drive.google.com/file/d/1s1PuYacaWC3XRvly-TrqybCXMUtseZ2q/view?usp=share_link)\n",
    "\n",
    "We have renamed those files using the prefix 'itgdbmock'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f08954a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ete3 env\n",
    "from ete3 import NCBITaxa\n",
    "import pickle\n",
    "from os.path import join, basename, exists\n",
    "import update_taxonomy\n",
    "from re import search\n",
    "import utils\n",
    "import pandas as pd\n",
    "import random\n",
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6215d7c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mockrobiota_src = 'mockrobiota/src'\n",
    "itgdbmock_files = {\n",
    "    'seq':join(mockrobiota_src, 'itgdbmock_seqs.fasta'),\n",
    "    'taxa':join(mockrobiota_src, 'itgdbmock_taxa.txt')}\n",
    "\n",
    "# Output files\n",
    "mockrobiota_files = {\n",
    "    'seq':join(mockrobiota_src, 'mockrobiota_seqs.fasta'),\n",
    "    'taxa':join(mockrobiota_src, 'mockrobiota_taxa.txt')}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e224c5b",
   "metadata": {},
   "source": [
    "## Filter mockrobiota records"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2f5f04a",
   "metadata": {},
   "source": [
    "From the itgdbmock we'll only keep the record corresponding to mockrobiota, as we do not trust the taxonomy assigment performed on PacBio HMP and Zymo mock-communities (using Blast) because we think that could be biased."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "806f89bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Taxa\n",
    "!grep 'Mockrobiota*\\|Feature ID' {itgdbmock_files['taxa']} > {mockrobiota_files['taxa']}\n",
    "# Seq\n",
    "!seqkit grep -p 'Mockrobiota*' -r {itgdbmock_files['seq']} > {mockrobiota_files['seq']}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0142445",
   "metadata": {},
   "source": [
    "## Format taxonomy file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8d70a6f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load\n",
    "db_mockro = utils.load_db_from_files(taxa_path = mockrobiota_files['taxa'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ade9f4a0",
   "metadata": {},
   "source": [
    "Change species format according SILVA species format (genus + specie)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "aea83317",
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in db_mockro.values():\n",
    "    # change 'Chlorobium' genus for 'Pelodictyon' according \n",
    "    #+ to NCBI taxonomy\n",
    "    record['taxa'][5] = record['taxa'][5].replace(\n",
    "        'Chlorobium', 'Pelodictyon')\n",
    "    # change species format\n",
    "    record['taxa'][6] = f\"{record['taxa'][5]}_{record['taxa'][6]}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc47c352",
   "metadata": {},
   "source": [
    "Update taxonomy acoording to NCBI using ete3 module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f24c0764",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original size of Mockrobiota: 274\n",
      "Final size of Mockrobiota: 274\n",
      "Total found taxa: 274\n",
      "60.94890510948905 % of the taxa changed and was updated\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%autoreload\n",
    "\n",
    "not_found_taxa = []\n",
    "\n",
    "key = 'Mockrobiota'\n",
    "\n",
    "# Information\n",
    "found_taxa = 0\n",
    "change_taxa = 0\n",
    "len_original_db = len(db_mockro)\n",
    "print(f\"Original size of {key}: {len_original_db}\")\n",
    "\n",
    "# Copy database\n",
    "db_mockro_copy = db_mockro.copy()\n",
    "\n",
    "\n",
    "for ID, items in db_mockro.items():\n",
    "    taxa = items['taxa']\n",
    "    new_taxa = update_taxonomy.update_taxonomy_ncbi(taxa, join_taxa = False)\n",
    "\n",
    "\n",
    "    if new_taxa:\n",
    "        # check if taxa was changed\n",
    "        if new_taxa != taxa:\n",
    "            change_taxa +=1\n",
    "\n",
    "\n",
    "        # check if there is still species level\n",
    "        if new_taxa[6]:\n",
    "            # replace taxa and count it\n",
    "            db_mockro[ID]['taxa'] = new_taxa\n",
    "            found_taxa = found_taxa + 1 \n",
    "        else:\n",
    "            db_mockro_copy.pop(ID)\n",
    "\n",
    "    else:\n",
    "        not_found_taxa.append(taxa)\n",
    "\n",
    "db_mockro = db_mockro_copy\n",
    "\n",
    "# Information\n",
    "len_db = len(db_mockro_copy)\n",
    "print(f\"Final size of {key}: {len_db}\")\n",
    "print(f\"Total found taxa: {found_taxa}\")\n",
    "print(f\"{(change_taxa/len_db)*100} % of the taxa changed and was updated\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32a55571",
   "metadata": {},
   "source": [
    "Downloand updated taxonomy file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ebb91ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "utils.download_db_from_dict(db_mockro, mockrobiota_files['taxa'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d427bd8c",
   "metadata": {},
   "source": [
    "## Create abundance_env.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "829cd392",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "274"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load\n",
    "db_mockro = utils.load_db_from_files(taxa_path = mockrobiota_files['taxa'])\n",
    "len(db_mockro)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61497d8",
   "metadata": {},
   "source": [
    "#### Approach 2\n",
    "\n",
    "- subset 200 entries for each sample (no repetition)\n",
    "- create feature table and abundance table. \n",
    "- create environment taxa and fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8193603",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mockro_ids = list(db_mockro.keys())\n",
    "\n",
    "for i in range(0, 5):\n",
    "    chosen_ids = random.sample(mockro_ids, k=200)\n",
    "    all_taxonomies = [\n",
    "    utils.join_taxa_lineages(record['taxa']\n",
    "                            ) for ID, record in db_mockro.items()\n",
    "        if ID in chosen_ids\n",
    "]\n",
    "    colname= f\"mockrobiota0{i}\"\n",
    "    \n",
    "    if i == 0:\n",
    "        abun_tab = pd.Series(all_taxonomies).value_counts().to_frame(\n",
    "        ).rename(columns={0: colname})\n",
    "        feature_tab = pd.Series(chosen_ids).value_counts().to_frame(\n",
    "        ).rename(columns={0: colname})\n",
    "        \n",
    "    else:\n",
    "        colname= f\"mockrobiota0{i}\"\n",
    "        abun = pd.Series(all_taxonomies).value_counts().to_frame(\n",
    "        ).rename(columns={0: colname})\n",
    "        feat = pd.Series(chosen_ids).value_counts().to_frame(\n",
    "        ).rename(columns={0: colname})\n",
    "        \n",
    "        \n",
    "        abun_tab = abun_tab.join(abun, how='outer').fillna(0)\n",
    "        feature_tab = feature_tab.join(feat, how='outer').fillna(0) \n",
    "        \n",
    "        \n",
    "    abun_tab[colname] = (abun_tab[colname]/abun_tab[colname].sum())  \n",
    "\n",
    "abun_tab = abun_tab.reset_index().rename(columns={'index': 'Taxon'})\n",
    "feature_tab = feature_tab.reset_index().rename(columns={'index': 'ID'})\n",
    "display(abun_tab.sort_values(by=\"mockrobiota00\", ascending=False))\n",
    "display(feature_tab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "821d1ada",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the table\n",
    "force=True\n",
    "\n",
    "exp_tax_mockro = join(mockrobiota_src,\"expected-taxonomy.tsv\")\n",
    "feature_tab_mockro = join(mockrobiota_src, \"expected-featuretab.tsv\")\n",
    "\n",
    "if not exists(exp_tax_mockro) or force:\n",
    "    abun_tab.to_csv(exp_tax_mockro, sep=\"\\t\", index=False)\n",
    "    \n",
    "if not exists(feature_tab_mockro) or force:\n",
    "    feature_tab.to_csv(feature_tab_mockro, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0263baf",
   "metadata": {},
   "source": [
    "#### Approach 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f09fe7a9",
   "metadata": {},
   "source": [
    "Calculate the abundance of the original fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f3778f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add prefixes and join taxa into string and store in a list\n",
    "all_taxonomies = [\n",
    "    utils.join_taxa_lineages(record['taxa']\n",
    "                            ) for ID, record in db_mockro.items()\n",
    "]\n",
    "\n",
    "abun_tab = pd.Series(all_taxonomies\n",
    "  ).value_counts().to_frame().reset_index().rename(\n",
    "  columns={'index': 'Taxon', 0: f\"mockrobiota00\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7dcde34b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>mockrobiota00</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>k__Bacteria; p__Chloroflexi; c__Anaerolineae; ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>k__Bacteria; p__Actinobacteria; c__Actinomycet...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>k__Bacteria; p__Actinobacteria; c__Actinomycet...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Taxon  mockrobiota00\n",
       "86  k__Bacteria; p__Chloroflexi; c__Anaerolineae; ...              1\n",
       "87  k__Bacteria; p__Actinobacteria; c__Actinomycet...              1\n",
       "88  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...              1\n",
       "89  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...              1\n",
       "90  k__Bacteria; p__Actinobacteria; c__Actinomycet...              1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abun_tab.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbabc977",
   "metadata": {},
   "source": [
    "Now, create 5 different samples with random abundancies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "206fa55c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>mockrobiota00</th>\n",
       "      <th>mockrobiota01</th>\n",
       "      <th>mockrobiota02</th>\n",
       "      <th>mockrobiota03</th>\n",
       "      <th>mockrobiota04</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.202899</td>\n",
       "      <td>0.022222</td>\n",
       "      <td>0.359551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.216216</td>\n",
       "      <td>0.115942</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>0.044944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.098765</td>\n",
       "      <td>0.108108</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.134831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.216216</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.088889</td>\n",
       "      <td>0.134831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Bacilli; o__Bac...</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028986</td>\n",
       "      <td>0.155556</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Taxon  mockrobiota00  \\\n",
       "0  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...       0.000000   \n",
       "1  k__Bacteria; p__Proteobacteria; c__Gammaproteo...       0.000000   \n",
       "2  k__Bacteria; p__Firmicutes; c__Clostridia; o__...       0.098765   \n",
       "3  k__Bacteria; p__Firmicutes; c__Clostridia; o__...       0.024691   \n",
       "4  k__Bacteria; p__Firmicutes; c__Bacilli; o__Bac...       0.222222   \n",
       "\n",
       "   mockrobiota01  mockrobiota02  mockrobiota03  mockrobiota04  \n",
       "0       0.027027       0.202899       0.022222       0.359551  \n",
       "1       0.216216       0.115942       0.266667       0.044944  \n",
       "2       0.108108       0.086957       0.000000       0.134831  \n",
       "3       0.216216       0.000000       0.088889       0.134831  \n",
       "4       0.000000       0.028986       0.155556       0.000000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ncols = 5 \n",
    "\n",
    "random.seed(123)\n",
    "\n",
    "rows = [random.choices(range(0, abun), k=ncols) for abun in abun_tab['mockrobiota00']]\n",
    "\n",
    "# Create columns\n",
    "for col in range(1,5):\n",
    "    abun_tab[f\"mockrobiota0{col}\"] = 0\n",
    "# Assign values \n",
    "for i in abun_tab.index:\n",
    "    abun_tab.loc[i,'mockrobiota0':] = rows[i]\n",
    "\n",
    "for col in range(0,5):\n",
    "    colname= f\"mockrobiota0{col}\"\n",
    "    # Keep abundance in relative\n",
    "    abun_tab[colname] = (abun_tab[colname]/abun_tab[colname].sum())*2\n",
    "\n",
    "display(abun_tab.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "abaa9772",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the table\n",
    "force=True\n",
    "\n",
    "exp_tax_mockro = join(mockrobiota_src,\"expected-taxonomy.tsv\")\n",
    "\n",
    "if not exists(exp_tax_mockro) or force:\n",
    "    abun_tab.to_csv(exp_tax_mockro, sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete3]",
   "language": "python",
   "name": "conda-env-ete3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "401px",
    "width": "345px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
